### PR TITLE
Fzf integration

### PR DESCRIPTION
**Pull request by mg979**

### CODE REVIEW

File: autoload/fzf/bookmarks.vim [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        return a:t[:(a:n-1)]."…"
+    else
+        let spaces = a:n - len(a:t)
+        let spaces = printf("%".spaces."s", "")
+        return a:t.spaces
+    endif
+endfunction
+
+function! s:is_annotation(t)
+    if a:t =~ "^Annotation: " | return s:green("§ ".a:t[12:]) | endif
```
Comment (mg979) [inline autoload/fzf/bookmarks.vim]: I'm removing `"Annotation: "` to only show the annotation string, and marking annotations with the symbol `"§"`. If you prefer another symbol, you can replace it (an unicode icon could not be typed, so there would be no easy way to filter annotations). If you change the length of the symbol (by using `"!!! "`, for example) you should also change below to keep the alignment.

File: autoload/fzf/bookmarks.vim [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    let text = substitute(text, "^\t*", "", "")
+
+    " when colorizing text, the actual string will be longer than the one
+    " that is showed, store the difference and add it to the padding
+    let slen = len(text)
+    let text = substitute(text, "\t", s:cyan("\\\\t"), "g")
+    let diff = len(text) - slen
+
+    let annotation = s:is_annotation(text) | let slen = len(text)
+    if !empty(annotation) | let text = annotation | let diff = len(text) - slen + 10 | endif
```
Comment (mg979) [inline autoload/fzf/bookmarks.vim]: If you increase the length of the symbol, decrease the number 10 accordingly.

File: plugin/bookmark.vim [hunk 3]
```diff
@@ -32,6 +32,7 @@ call s:set('g:bookmark_auto_close',           0 )
 call s:set('g:bookmark_center',               0 )
 call s:set('g:bookmark_location_list',        0 )
 call s:set('g:bookmark_disable_ctrlp',        0 )
+call s:set('g:bookmark_prefer_fzf',           1 )
```
Comment (mg979) [inline plugin/bookmark.vim]: This could default to `0` if you prefer.

File: plugin/bookmark.vim [hunk 4]
```diff
@@ -302,6 +307,11 @@ command! -nargs=? BookmarkMoveUp call s:move_relative(<q-args>, -1)
 command! -nargs=? BookmarkMoveDown call s:move_relative(<q-args>, 1)
 command! -nargs=? BookmarkMoveToLine call s:move_absolute(<q-args>)
 
+command! -bang -nargs=? -complete=buffer FzfBookmarks call fzf#vim#ag(<q-args>, {
+            \ 'source': fzf#bookmarks#list(),
+            \ 'sink': function('fzf#bookmarks#open'), 'down': '30%',
+            \ 'options': '--prompt "Bookmarks  >>>  "'})
```
Comment (pyrho) [inline plugin/bookmark.vim]: Thanks for this PR!
Do you how to make this create a preview window within the FZF split ? I've tried various combinations to no avail.

Comment (mg979) [reply plugin/bookmark.vim]: I could be wrong, but I don't think it's possible to pass the selected line to a vim function. Fzf seems to pass it to the shell as `{}`, as in `--preview "echo {}"`.

Maybe passing it to a shell script (something like `--preview "sh vim_fzf_bookmarks.sh {}"`) that would parse the line, extract filename and line number, then use `sed` to show around those lines (like described [here](https://serverfault.com/questions/101900/read-specified-range-of-lines-from-a-file)).

Not sure where the shell script would go, though, or how to make it work under Windows (if it worked on Linux/Mac).

Comment (pyrho) [reply plugin/bookmark.vim]: ftr I've opened [a PR](https://github.com/mg979/vim-bookmarks/pull/1) on @mg979 's fork to deal with FZF previews.

**Comment by kidonchu** [comment]
@mg979 This PR will be incredibly useful. Do you know when this will get merged?

**Comment by mg979** [comment]
@kidonchu It's not up to me to decide. Are you using it? Any problems found with it?

**Comment by kidonchu** [comment]
@mg979 I've just pulled in your changes in this PR and started using it today. I will leave a comment if any issue is found.

**Comment by mg979** [comment]
Thanks, honestly I'm not using it much, so some extra testing is better.

**Comment by reaysawa** [comment]
I have a few objections with the direction taken here.

First is the reliance on ag, which I did not have installed and is not a requirement for this command to work.

Second is that you're wrapping FZF options under variables and passing down to the command, which is limiting in my opinion. And to worsen this problem, you're implementing your own sink for the command, so the user can't truly modify anything that happens outside of the behavior you chose to have implemented.

My suggestion is to provide a sample configuration for the commands and implement only the source, with a few more options if needed, then let people configure the command however they prefer. This seems to be more aligned with what the FZF has to offer.

I'm not using any of the FZF-specific variables to achieve what I want here and the result is the same.

```
let g:fzf_action = {
  \ 'ctrl-q': function('s:fzf_build_quickfix'),
  \ 'ctrl-t': 'tab split',
  \ 'ctrl-s': 'split',
  \ 'ctrl-v': 'vsplit' }

fun! s:fzf_bookmarks_sink_star(action_and_line)
  let [action, line] = a:action_and_line
  let [filename, line_nr] = matchlist(line, '^\(.*\):\([0-9]\+\)')[1:2]
  execute get(g:fzf_action, action, 'e').' '.filename
  execute 'call cursor(['.line_nr.',1])'
endfun

fun! s:fzf_bookmarks()
  let opts = ['--ansi', '--multi',
              \ '--prompt', 'Bookmarks > ',
              \ '--color', 'hl:4,hl+:12',
              \ '--preview-window', 'up:50%',
              \ '--preview', shellescape(expand('$HOME/.plugged/fzf.vim/bin/preview.rb')).' {}',
              \ '--expect='.join(keys(get(g:, 'fzf_action')), ',')]
  call fzf#run(fzf#wrap({
        \ 'source': fzf#bookmarks#list(1),
        \ 'down': '60%',
        \ 'sink*': function('s:fzf_bookmarks_sink_star'),
        \ 'options': opts}))
endfun

command! MyFzfBookmarks call <SID>fzf_bookmarks()
nnoremap <silent> <Leader>b :MyFzfBookmarks<CR>
```

For instance, I'm choosing not to have arguments fed into the function here. I can also control the colors and the bindings however I want here. I think having the sample implementation as a starting points benefits both experienced and new users, plus it's one less thing the maintainer will have to deal with.

**Comment by mg979** [comment]
Thanks for having a look at this, I don't use it so my interest in it is gone.

> First is the reliance on ag, which I did not have installed and is not a requirement for this command to work.

Yes.. that should be changed. I'm not really practical with fzf.vim so it's probably more complicated than it should.

> 
> Second is that you're wrapping FZF options under variables and passing down to the command, which is limiting in my opinion. And to worsen this problem, you're implementing your own sink for the command, so the user can't truly modify anything that happens outside of the behavior you chose to have done.

For this kind of things, I prefer having a default behaviour that works more or less, and the possibility to customize it as you say. I guess your setup works already? So you can use `fzf#bookmarks#list(1)` and make your own command already. Is it a problem of documentation or something else?

Anyway I like the idea of using `g:fzf_action` to open the bookmark in a split, it could be made the default behaviour. Would you care at making a PR? Thanks

**Comment by reaysawa** [comment]
@mg979 Yes, the command does work already. Thanks for making the source!

The idea I was trying to bring up was to **not** define the default command inside of your changes directly, but to show the example FZF code in the README and let users define their own command using that as a guide (or just copy and paste into their own config). That would also mean to get rid of the variables as they would no longer be useful (except for **prefer_fzf**).

If that's fine with you I can try to do some cleanup and open the PR later. And if not you can just take the code I pasted above as an alternative in the documentation.

**Comment by mg979** [comment]
> The idea I was trying to bring up was to **not** define the default command inside of your changes directly, but to show the example FZF code in the README and let users define their own command

Ok but I like having a default behaviour that is good enough to be used. Why use fzf if you have to spend hours to set it up?

**Comment by jim-at-jibba** [comment]
I am not sure if this is still happening but I would love fzf integration. Thanks

**Comment by mg979** [comment]
It's pointless I keep a PR I can't follow because I don't use this plugin anymore.

**Pull request closed without merging**